aws/src/main/java/org/apache/iceberg/aws/AssumeRoleAwsClientFactory.java

```diff
@@ -43,17 +43,16 @@ public class AssumeRoleAwsClientFactory implements AwsClientFactory {
   private Set<Tag> tags;
   private int timeout;
   private String region;
-  private String s3Endpoint;
-  private boolean s3UseArnRegionEnabled;
-  private String dynamoDbEndpoint;
-  private String httpClientType;
+  private AwsProperties awsProperties;
+  private AssumeRoleRequest assumeRoleRequest;
 
   @Override
   public S3Client s3() {
     return S3Client.builder()
         .applyMutation(this::configure)
-        .applyMutation(builder -> AwsClientFactories.configureEndpoint(builder, s3Endpoint))
-        .serviceConfiguration(s -> s.useArnRegionEnabled(s3UseArnRegionEnabled).build())
+        .applyMutation(
+            builder -> AwsClientFactories.configureEndpoint(builder, awsProperties.s3Endpoint()))
```

Comment:
this method can also be moved to `awsProperties::applyEndpointConfiguration`, and we can set the endpoint based on the instance type, for example `if (builder instanceof S3ClientBuilder) { ... }`